### DISCUSSION TITLE
Hot Reloading / Web Sockets

### DISCUSSION THREAD

**Discussion by wardy484** [livewire/livewire#3474] [3. Feature ideas]
This is very much a query rather than an issue. @calebporzio I saw on your blog an early version used web sockets as the transfer by default and featured hot reloading (very nice). 

I've noticed that web sockets seems to have been removed in favour of http; I was just curious what lead to that decision. When I was the framework I thought oh this is kind of cool but it seems like if you used web sockets you'd eliminate the overhead of opening a new http connection for every state change. So I'm curious what lead to the current implementation? 

Very nice work though; I definitely plan on tinkering around with livewire.

**Comment by davidpiesse** [comment]
Hi @wardy484 
I know @calebporzio will likely jump in on this also but I can just about remember some of the reasoning.

Initially it was Web Sockets and Caleb built it as such. The issue was around having to install some web socket capability (not too onerous - but an extra dependancy). It also meant the number of users would directly correlate to the number of sockets open and thus resource usage on your machine. This would quickly bottleneck on a server if you went viral!

So in comes HTTP. We now have the non-persistent connection but we can handle way more requests from more users than we can open web socket connections. It is also a known technology used in this way that scales better. 
Caleb saw Github do it in this way and didn't look back. 

There is still the ability to use an alternative driver to communicate between the front and back end, and someone could write a web socket alternative if they so wanted.

The one place web sockets are (potentially) used is with the eventing and Laravel Echo. This still uses all the default ways of communicating with the backend (often Pusher / laravel-websockets).

David :)

**Comment by wardy484** [comment]
Hi @davidpiesse,

That's really insightful I had suspicions it was something to do with ease of use or lowering the barrier to entry but wasn't sure. Do you know of any articles describing github doing it this way? I'd be interested to read more.

**Comment by davidpiesse** [comment]
Not that I am aware of - I just opened up Dev Tools, and started clicking on drop downs etc.,  and seeing the network activity.
These may be of help https://github.blog/2018-09-06-removing-jquery-from-github-frontend/
but it's a bit tricky trying to google GitHub 'app' info with the wealth of user code a search brings up! If you find anything else add it here! :)

**Comment by davidpiesse** [comment]
This is how GitHub do it I believe...
https://github.com/github/include-fragment-element

**Comment by calebporzio** [comment]
What @davidpiesse said lol  - thanks David!

**Comment by wardy484** [comment]
Thanks for the response. That was quite insight.

**Comment by bgarrison25** [comment]
I know this is closed but I was considering the exact same thing. Those are some good reasons and I think its a really great idea. I'm really looking forward to what you guys do with this.

**Comment by davidpiesse** [comment]
I can see a next step that goes towards project-x in a web component thing that Livewire can tap into as well..

**Comment by calebporzio** [comment]
Please explain your grand vision Mr. Piesse (@davidpiesse). I must know.

**Comment by davidpiesse** [comment]
Well I just spent 20 mins writing a response, and just deleted it! 

My idea has some/many holes that need thinking/ mulling / some river time to consider.

Some parts are no brainers (using project-x to replace a lot of the Livewire JS code) but the component-izing of a Livewire component isn't there yet...

🤔🤔🤔

**Comment by calebporzio** [comment]
Ah yes, I definitely think there is something there too, just don't see it quite yet.

Definitely put some river time in and get back to me!

**Comment by valeriu-dev** [comment]
Maybe we should re-open this issue?

https://alistapart.com/article/the-future-of-web-software-is-html-over-websockets/

**Reply by joshhanley** [reply]
@vodnicearv can you summarise the article and why you think it's important?

Livewire was originally built on websockets and a decision was made to go away from it. But Livewire supports websockets for listening to events. https://laravel-livewire.com/docs/2.x/laravel-echo

How do you propose it should be?

Thanks!

**Reply by valeriu-dev** [reply]
@joshhanley I have some difficulty in explaining things in English. But I'll give it a try.
Web sockets have the advantage that they are faster because they are always open.
But that doesn't mean you have to give up http. It's nice to have a choice.

Can we add sockets as an additional data transport protocol?

The connection would only be between the client who opened the site and the server. If the socket tab is closed, the connection will also be closed.

**Reply by joshhanley** [reply]
@vodnicearv no problem, I think you explained it well 🙂

Yep I think it would be nice to have a choice what protocol to use. I'm not sure how feasible that would be in the current version of Livewire. But would be a decent feature to add into V3.

**Reply by valeriu-dev** [reply]
@joshhanley nice :)

I'm interested what Caleb @calebporzio thinks about this, now

**Reply by valeriu-dev** [reply]
https://calebporzio.com/livewire-isnt-actually-live

**Reply by joshhanley** [reply]
@vodnicearv that doesn't mean it can't be though 😉

**Comment by joshhanley** [comment]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/livewire/livewire/discussions/3147
